docs/spec.md

```diff
@@ -0,0 +1,269 @@
+- We let the resource try and handle the operation. For example, `A1 + A2` effectively desugars to:
+  - `context.request(host, "A1").request(host, "ADD", context.request(host, "A2"))`
+  - In effect, we call the `ADD` method of `A1` with argument `A2`.
+- We have a hybrid approach: First try and detect an overload, otherwise fallback to `asValue`.
+
+##### Compilation
+
+Formulas should compile to JavaScript functions of the type:
+
+```ts
+(host: Host, context: Resource) => CalcValue;
+```
+
+[JW-NOTE: We might want to encode run-time errors here, so the result type is actually `Result<CalcValue, Error>`. Some thought is needed on this. It can be pretty robust but also verbose to deal with. If we make `Result` a `Thenable` interface then we can easily have asynchronous formulas of the type `(host: Host, context: Resource) => Promise<CalcValue>;`]
+
+Example 1, assuming synchronous evaluation:
+
+- `=Table1.ColumnA.Sum` compiles to
+- `(context, host) => context.request(host, 'Table1', id).request(host, 'ColumnA', id).request(host, 'Sum', id)`
+
+Example 2, assuming synchronous evaluation:
+
+- `=Table1.ColumnA.Sum + Table1.ColumnB.Max` compiles to
+
+```ts
+(context, host) =>
+  calc.add(
+    context
+      .request(host, "Table1", id)
+      .request(host, "ColumnA", id)
+      .request(host, "Sum", id),
+    context
+      .request(host, "Table1", id)
+      .request(host, "ColumnB", id)
+      .request(host, "Max", id)
+  );
+```
+
+where `calc.add` is the `+` operator with the following semantics. Given `calc.add(a,b)`
+
+- If `a` evaluates to a resource handle `add` using the extensibility model we define.
+- If `b` evaluates to a resource handle `add` using the extensibility model we define (assuming `a` is not a resource).
+- Otherwise, return `a + b`.
+
+[JW-NOTE: What happens if `a` or `b` are errors?]
+
+##### Efficient compilation
+
+The example `=Table1.ColumnA.Sum + Table1.ColumnB.Max` compiles to (somewhat) verbose code. This gives us a great deal of flexibility but may often be too general and impact performance. In some cases the ideal code to generate would be
+
+```ts
+(context, host) => context.Table1.ColumnA.Sum + context.Table1.ColumnB.Max;
+```
+
+where we can actually benefit from inlining caching, and we use the raw `+` operator. This is especially true if we use resources that effectively represent records, or JavaScript objects with fixed properties. To be able to support this type of compilation one proposal is the notion of _resource schemas_.
+
+A resource schema is a static description of a resource (or its type) that is attached to the resource and available during parse. A schema for our context resource might look like:
+
+```ts
+const contextSchema = {
+  Table1: {
+    ColumnA: {
+      Sum: "number",
+      Max: "number"
+    },
+    ColumnB: {
+      Sum: "number",
+      Max: "number"
+    }
+  }
+};
+```
+
```

Comment:
(Similar to Intellisense, this schema would probably need to be something the parser queries on demand for individual properties.)